clients/src/main/java/org/apache/kafka/clients/consumer/internals/StreamsMembershipManager.java

```diff
@@ -0,0 +1,823 @@
+            }
+        }
+
+        Optional<LocalAssignment> updateWith(final Map<String, SortedSet<Integer>> activeTasks,
+                                             final Map<String, SortedSet<Integer>> standbyTasks,
+                                             final Map<String, SortedSet<Integer>> warmupTasks) {
+            if (localEpoch != NONE_EPOCH) {
+                if (activeTasks.equals(this.activeTasks) &&
+                    standbyTasks.equals(this.standbyTasks) &&
+                    warmupTasks.equals(this.warmupTasks)) {
+
+                    return Optional.empty();
+                }
+            }
+
+            long nextLocalEpoch = localEpoch + 1;
+            return Optional.of(new LocalAssignment(nextLocalEpoch, activeTasks, standbyTasks, warmupTasks));
+        }
+
+        @Override
+        public String toString() {
+            return "LocalAssignment{" +
+                "localEpoch=" + localEpoch +
+                ", activeTasks=" + activeTasks +
+                ", standbyTasks=" + standbyTasks +
+                ", warmupTasks=" + warmupTasks +
+                '}';
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            LocalAssignment that = (LocalAssignment) o;
+            return localEpoch == that.localEpoch &&
+                Objects.equals(activeTasks, that.activeTasks) &&
+                Objects.equals(standbyTasks, that.standbyTasks) &&
+                Objects.equals(warmupTasks, that.warmupTasks);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(localEpoch, activeTasks, standbyTasks, warmupTasks);
+        }
+    }
+
+    static final Utils.TopicPartitionComparator TOPIC_PARTITION_COMPARATOR = new Utils.TopicPartitionComparator();
+
+    private final Logger log;
+
+    private final StreamsAssignmentInterface streamsAssignmentInterface;
+
+    private final SubscriptionState subscriptionState;
+
+    private final ConsumerMetadata consumerMetadata;
+
+    private final BackgroundEventHandler backgroundEventHandler;
+
+    private final Map<String, Uuid> assignedTopicIdCache = new HashMap<>();
+
+    private MemberState state;
+
+    private final String groupId;
+
+    private String memberId = "";
+
+    private final Optional<String> groupInstanceId = Optional.empty();
+
+    private int memberEpoch = 0;
+
+    private Optional<CompletableFuture<Void>> leaveGroupInProgress = Optional.empty();
+
+    private CompletableFuture<Void> staleMemberAssignmentRelease;
+
+    private boolean reconciliationInProgress;
+
+    private boolean rejoinedWhileReconciliationInProgress;
+
+    private final Optional<ClientTelemetryReporter> clientTelemetryReporter;
+
+    private LocalAssignment targetAssignment = LocalAssignment.NONE;
```

Comment:
This is called "currentTargetAssignment" in `AbstractMembershipManager`. Why the different names?

Reply:
Because `currentTargetAssignment` was confusing. We had a `currentAssignment` and a `currentTargetAssignment`. The `current` prefix did contain any information. The important aspect is that one is the assignment currently installed, i.e, `currentAssignment` and the other is the assignment we want to reach, i.e., `targetAssignment`.